---
title: "Software"
permalink: /software/
author_profile: true
---

Here are some of the software products to come out of my research:

### SpatialCorr

Identify gene sets with spatially varying correlation structure.

* GitHub: [https://github.com/mbernste/SpatialCorr](https://github.com/mbernste/SpatialCorr)
* Documentation: [https://spatialcorr.readthedocs.io/en/latest/index.html](https://spatialcorr.readthedocs.io/en/latest/index.html)

### CellO

Cell type classification of single-cell RNA-seq data against the Cell Ontology.

* GitHub: [https://github.com/deweylab/CellO](https://github.com/deweylab/CellO)

### CellO Viewer

Explore cell type-specific gene expression signatures across the Cell Ontology.

* Website: [https://uwgraphics.github.io/CellOViewer/](https://uwgraphics.github.io/CellOViewer/)
* GitHub: [https://github.com/uwgraphics/CellOViewer](https://github.com/uwgraphics/CellOViewer)

### MetaSRA Website

Search for public RNA-seq data within the Sequence Read Archive using biomedical ontologies.

* Website: [http://metasra.biostat.wisc.edu](http://metasra.biostat.wisc.edu)
* GitHub: 
  * [https://github.com/deweylab/MetaSRA-website-frontend](https://github.com/deweylab/MetaSRA-website-frontend)
  * [https://github.com/deweylab/MetaSRA-API-backend](https://github.com/deweylab/MetaSRA-API-backend)
 
### MetaSRA Pipeline

Standardize biomedical metadata that are encoded as key-value pairs.

* GitHub: [https://github.com/deweylab/MetaSRA-pipeline](https://github.com/deweylab/MetaSRA-pipeline)

### CHARTS

Characterize and compare tumor subpopulations across public single-cell RNA-seq data.

* Website: [https://charts.morgridge.org](https://charts.morgridge.org)
* GitHub: [https://github.com/stewart-lab/CHARTS](https://github.com/stewart-lab/CHARTS)

### Series Finder & Case-Control Finder

Build structured datasets of RNA-seq data from the Sequence Read Archive.

* GitHub: [https://github.com/mbernste/hypothesis-driven-SRA-queries](https://github.com/mbernste/hypothesis-driven-SRA-queries)

